js/packages/teams-ai/src/prompts/PromptSectionBase.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * @module teams-ai
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Message } from "./Message";
+import { PromptFunctions } from "./PromptFunctions";
+import { PromptSection, RenderedPromptSection } from "./PromptSection";
+import { TurnContext } from 'botbuilder';
+import { Tokenizer } from "../tokenizers";
+import { Memory } from "../MemoryFork";
+
+/**
+ * Abstract Base class for most prompt sections.
+ */
+export abstract class PromptSectionBase implements PromptSection {
+    public readonly required: boolean;
+    public readonly tokens: number;
+    public readonly separator: string;
+    public readonly textPrefix: string;
+
+    /**
+     * Creates a new 'PromptSectionBase' instance.
+     * @param tokens Optional. Sizing strategy for this section. Defaults to `auto`.
+     * @param required Optional. Indicates if this section is required. Defaults to `true`.
+     * @param separator Optional. Separator to use between sections when rendering as text. Defaults to `\n`.
+     * @param textPrefix Optional. Prefix to use for text output. Defaults to `undefined`.
+     */
+    public constructor(tokens: number = -1, required: boolean = true, separator: string = '\n', textPrefix: string = '') {
+        this.required = required;
+        this.tokens = tokens;
+        this.separator = separator;
+        this.textPrefix = textPrefix;
+    }
+
+    public async renderAsText(context: TurnContext, memory: Memory, functions: PromptFunctions, tokenizer: Tokenizer, maxTokens: number): Promise<RenderedPromptSection<string>> {
+        // Render as messages
+        const asMessages = await this.renderAsMessages(context, memory, functions, tokenizer, maxTokens);
+        if (asMessages.output.length === 0) {
+            return { output: '', length: 0, tooLong: false };
+        }
+
+        // Convert to text
+        let text = asMessages.output.map((message) => PromptSectionBase.getMessageText(message)).join(this.separator);
+
+        // Calculate length
+        const prefixLength = tokenizer.encode(this.textPrefix).length;
+        const separatorLength = tokenizer.encode(this.separator).length;
+        let length = prefixLength + asMessages.length + ((asMessages.output.length - 1) * separatorLength);
+
+        // Truncate if fixed length
+        text = this.textPrefix + text;
+        if (this.tokens > 1.0 && length > this.tokens) {
+            const encoded = tokenizer.encode(text);
```

Comment:
this logic looks similar to that of the below function `returnMessages`, maybe we can consolidate and reuse it here.